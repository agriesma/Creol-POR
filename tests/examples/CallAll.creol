/*
 * CallAll.creol -- Shows how one call a variable number of objects
 * and syncrhonise on their return.
 *
 * This file is part of creoltools
 *
 * Written and Copyright (c) 2007 by Marcel Kyas
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

interface I
begin
  with Any op m(in i: Data; out o: Data)
end

class CallAll
begin
  var outstanding: Int
  var list: List[[I, Data]]
  var results: List[[I, Data]]

  op doit(in l: List[[I, Data]]) ==
    var i: [I, Data];
    var r: Label[Data];
    var o:Data;
    if l /= nil then
      r!fst(i).m(snd(i)); 
      doit(tail(l); );
      await r?;
      r?(o);
      results := results |- (fst(i), o)
    end

  op call_all ==
     results := nil;

     /* The next call must be synchronous to assert that outstanding is 0,
        if and only if all called objects have replied. */

    doit(list; );

    await length(results) = length(results)

end
