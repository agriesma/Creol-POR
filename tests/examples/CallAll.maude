load creol-interpreter .
mod PROGRAM is
protecting CREOL-SIMULATOR .
op classes : -> Configuration [ctor] .
eq classes =
< "CallAll" : Cl | Inh: (noInh),
  Par: (noVid),
  Att: ("outstanding" |-> null , "list" |-> null , "results" |-> null),
  Mtds: (< "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
    < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
    < "doit" : Mtdname | Param: "l", Latt: "l" |-> null , "label:0" |-> null , "i" |-> null , "r" |-> null , "o" |-> null , "_" |-> null, Code: if "/=" ( "l" :: list(emp) ) th "r" ! "fst" ( "i" ) . "m" ( "snd" ( "i" ) ) ; ( "label:0" ! "this" . "doit" ("tail" ( "l" ) ) )  ; ( "label:0" ? ( noVid ) )  ; ( await ( "r" ?? ) ) ; ( "r" ? ( "o" ) )  ; "results" ::= "|-" ( "results" :: pair("fst" ( "i" ) , "o") ) el skip fi ; return ( emp ) > *
    < "call_all" : Mtdname | Param: noVid, Latt: "label:0" |-> null , "_" |-> null, Code: "results" ::= list(emp) ; ( "label:0" ! "this" . "doit" ("list" ) )  ; ( "label:0" ? ( noVid ) )  ; ( await "=" ( "#" ( "results" ) :: "#" ( "list" ) ) ) ; return ( emp ) >),
  Ocnt: 0 >

.
endm
