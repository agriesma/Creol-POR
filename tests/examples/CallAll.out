==========================================
reduce in PROGRAM : classes .
result Class: 
< "CallAll" : Cl | Inh: noInh, Par: noVid, Att: "list" |-> null, "outstanding"
    |-> null, "results" |-> null, Mtds:
  < "call_all" : Mtdname | Param: noVid, Latt: "_" |-> null, "label:0" |->
    null, Code: "results" ::= list(emp) ; "label:0" ! "this" . "doit"("list") ;
    "label:0" ?(noVid) ; await "="("#"("results") :: "#"("list")) ; return(emp)
    > *
  < "doit" : Mtdname | Param: "l", Latt: "_" |-> null, "i" |-> null, "l" |->
    null, "label:0" |-> null, "o" |-> null, "r" |-> null, Code: if "/="("l" ::
    list(emp)) th "r" ! "fst"("i") . "m"("snd"("i")) ; "label:0" ! "this" .
    "doit"("tail"("l")) ; "label:0" ?(noVid) ; await ("r" ??) ; "r" ?("o") ;
    "results" ::= "|-"("results" :: pair("fst"("i"), "o")) el skip fi ; return(
    emp) > *
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return(
    emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return(
    emp) >,
  Ocnt: 0 >

Bye.
